source/extensions/filters/common/rbac/matching/inputs.cc

```diff
@@ -0,0 +1,39 @@
+#include "source/extensions/filters/common/rbac/matching/inputs.h"
+
+#include "envoy/registry/registry.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RBAC {
+namespace Matching {
+
+using DestinationIPInputFactory = Network::Matching::DestinationIPInputBaseFactory<MatchingData>;
+using DestinationPortInputFactory =
+    Network::Matching::DestinationPortInputBaseFactory<MatchingData>;
+using SourceIPInputFactory = Network::Matching::SourceIPInputBaseFactory<MatchingData>;
+using SourcePortInputFactory = Network::Matching::SourcePortInputBaseFactory<MatchingData>;
+using DirectSourceIPInputFactory = Network::Matching::DirectSourceIPInputBaseFactory<MatchingData>;
+using ServerNameInputFactory = Network::Matching::ServerNameInputBaseFactory<MatchingData>;
+using SourceTypeInputFactory = Network::Matching::SourceTypeInputBaseFactory<MatchingData>;
+
+REGISTER_FACTORY(DestinationIPInputFactory, Matcher::DataInputFactory<MatchingData>);
```

Comment:
I was thinking that there is little difference between HTTP data input and RBAC data input for HTTP requests. Is there a reason why we want to duplicate registration bits?

Reply:
I am not sure what you mean, but there is no duplicate registration.

Reply:
I'd like more thoughts from @snowp here. I think it's redundant to have both HTTP matching data and HTTP RBAC matching data. The context for matching is essentially the same, and it costs us complexity in duplicate registrations for all input factories. The only benefit is that we can allow-list a subset of inputs (this probably should be done in a more principled way for all match trees). Maybe, I'm missing something?

Reply:
The RBAC matching data is shared for both network and HTTP RBAC filter. As described in #20324, we currently cannot support registration inheritance, and as a result, we must have a new matching data if it has different supported inputs.

The RBAC matching data supports not only network and HTTP inputs, but also metadata and authenticated (though not completed) inputs. So we have to add a new matching data, and have inputs registered.

Reply:
BTW, I think we should reopen #20324 and mark it help-wanted.

Reply:
My concern is that I'm not sure we need separate sets of inputs for RBAC vs HTTP as defined in code.

I think metadata matching makes sense for network listener matching as well, for example. There is a use case for using listener filters to extract metadata from the network socket options and then match on that. Also for the internal listeners, some context will be propagated from the downstream in a metadata form. So whenever metadata input comes in, we'll cross-register it for network, HTTP, and RBAC.

I suspect we do need controls on permitted types of inputs in RBAC matchers. I'm just not convinced creating separate static registrations (one for each RBAC, network, HTTP) is how we should do this.

Reply:
It is reasonable to reuse the HTTP matching data, and in that case, we can use different validation visitor to avoid unsupported inputs (like trailer and response inputs).

But it also comes another problem. If we want to introduce new inputs, we will have a much heavier global matching data, and the related extensions should also modify their validation visitor to allow or disallow the input.

So I think either approach is OK. If you think the former one looks better, I will change my code, and which will also this PR smaller (and I am also willing to see that).

Reply:
I think either way is reasonable, but it's important to make a decision now before we start changing the other filters. I am not very worried about matching data, it looks like a thin wrapper and can be reused.

So the two choices are:
1) Custom `MatchingData` per filter matcher tree - extensions explicitly register `Input` factories for the input.
2) Generic `HTTPMatchingData` for all HTTP filters - factories are registered once, but extensions explicitly list permitted types in validation visitors.

One complication is that RBAC has its own extensions (e.g. https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/rbac/matchers/upstream_ip_port/v3/upstream_ip_port_matcher.proto#extension-envoy-rbac-matchers-upstream-ip-port). So the option (2) cannot be a static list, it needs some "sub-category" or label on the factory.

I might be late to this discussion, and if this was already decided upon with @snowp, I'm happy with either direction.

Reply:
I think the way this was originally envision we'd have HTTP filters use the ExtensionWithMatcher and declare what could be matched on via https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/composite/config.h#L31-L44 etc. 

Originally I wanted to make the built in validator allow specifying a "phase" instead of a list of specific input extensions, e.g. "this filter needs an answer by the time it sees request headers" -> only allow inputs that can be evaluated before or with request headers. With that model any input extension would just naturally hook up to the HttpMatchingData to add more information which would be available to all HTTP filters that want to use matching, and the "phase" requirement would dictate which inputs would be permissable. So while RBAC might add their own matchers, if this is available early enough I don't see a particular reason to then disallow other filter to use it

Now this was never implemented, and I'm not quite sure how complicated something like this would be to do. Additionally, I think we've veered in the direction of filters managing their own match evaluation, which might make this harder to do in practice

In general I'd lean towards something that allows for the most reuse as possible, so I'd probably avoid having to define custom MatchingData types for each use case if possible

Reply:
Thanks. Agree on maximizing re-use across filters for now, and adding extra constraints on what inputs are available when later. FWIW network inputs are generally available early enough because of the "peeking" listener filters.